File: 2023.1-Matematica/lista4/B.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

ll powLL(ll base, ll exp)
{
    ll result = 1;
    for (;;)
    {
        if (exp & 1)
            result *= base;
        exp >>= 1;
        if (!exp)
            break;
        base *= base;
    }

    return result;
}

int main()
{
    ll p;
    cin >> p;
    ll total = 0;

    for (ll x = 1; x < p; x++)
    {
        if ((powLL(x, p - 1) - 1) % p != 0)
            continue;

        bool pr = true;
        for (ll exp = p - 2; exp > 0; exp--)
        {
            if ((powLL(x, exp) - 1) % p == 0)
            {
                pr = false;
                break;
            }
        }
        if (pr) total++;
    }
    cout << total << endl;

    return 0;
}

File: 2023.1-Matematica/lista4/B.py
def main():
    p = int(input())
    total = 0
    for x in range(1, p):
        if (x**(p - 1) - 1) % p != 0:
            continue

        for exp in range(p - 2, 0, -1):
            if (x**exp - 1) % p == 0:
                break
        else:
            total += 1

    print(total)

if __name__ == '__main__':
    main()
